ethereum/mock-p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/testing/MockNetwork.java

```diff
@@ -246,13 +247,14 @@ public PeerInfo getPeer() {
     }
 
     @Override
-    public void terminateConnection(final DisconnectReason reason, final boolean peerInitiated) {
+    public void terminateConnection(
+        final Optional<DisconnectReason> reason, final boolean peerInitiated) {
       network.disconnect(this, reason);
     }
 
     @Override
     public void disconnect(final DisconnectReason reason) {
-      network.disconnect(this, reason);
+      network.disconnect(this, Optional.ofNullable(reason));
```

Comment:
Should we push Optional<DisconnectReason> into PeerConnection?  An alternative would be to make a no-args disconnect() and make disconnect(@NotNull final DisconnectReason reason) and handle the optional stuff in the impls.

Reply:
I cleaned this up a little, but I have `disconnect()` and `disconnect(final DisconnectReason reason)` methods on `PeerConnection`.  I thought this was cleaner than having to wrap the reasons in `Optional`'s all over.